memory/iter.go

```diff
@@ -7,210 +7,174 @@ import (
+	hasItem bool
 }
 
-func (iter *dsPrefixIter) Valid() bool {
-	return validForPrefix(iter.curItem, iter.prefix)
-}
+func newPrefixIter(db *Datastore, prefix []byte, reverse bool, version uint64) *iterator {
+	start := prefix
+	end := bytesPrefixEnd(prefix)
 
-func validForPrefix(item *dsItem, prefix []byte) bool {
-	if item == nil {
-		return false
+	it := db.values.Iter()
+	var hasItem bool
+	if reverse {
+		hasItem = it.Last()
+	} else {
+		hasItem = it.First()
 	}
 
-	return bytes.HasPrefix(item.key, prefix) && !bytes.Equal(item.key, prefix)
-}
+	iter := &iterator{
+		db:      db,
+		version: version,
+		it:      it,
+		start:   start,
+		end:     end,
+		// A prefix iterator should not return a key exactly matching itself.
+		mayExactlyMatchStart: false,
+		reverse:              reverse,
+		hasItem:              hasItem,
+	}
 
-func (iter *dsPrefixIter) Next() {
-	if iter.it.Next() {
-		iter.loadLatestItem()
+	if reverse {
+		iter.Seek(end)
 	} else {
-		iter.curItem = nil
+		iter.Seek(start)
 	}
-}
 
-func (iter *dsPrefixIter) Key() []byte {
-	if iter.curItem != nil {
-		return iter.curItem.key
-	}
-	return nil
+	return iter
 }
 
-func (iter *dsPrefixIter) Value() ([]byte, error) {
-	if iter.curItem != nil {
-		return iter.curItem.val, nil
+func newRangeIter(db *Datastore, start, end []byte, reverse bool, version uint64) *iterator {
+	it := db.values.Iter()
+	var hasItem bool
+	if reverse {
+		hasItem = it.Last()
+	} else {
+		hasItem = it.First()
 	}
-	return nil, nil
-}
 
-func (iter *dsPrefixIter) Seek(key []byte) {
-	// get the correct initial version for the seek
-	// if there exists an exact match in keys, use the latest version
-	// of that key, otherwise, use the provided DB version
-	// TODO this could use some "peek" mechanic instead of a full lookup
-	version := iter.version
-	result := iter.db.get(key, version)
-	if result.key != nil && !result.isDeleted {
-		version = result.version
+	iter := &iterator{
+		db:                   db,
+		version:              version,
+		it:                   it,
+		start:                start,
+		end:                  end,
+		mayExactlyMatchStart: true,
+		reverse:              reverse,
+		hasItem:              hasItem,
 	}
 
-	if iter.it.Seek(dsItem{key: key, version: version}) {
+	if len(end) > 0 && reverse {
+		iter.Seek(end)
+	} else if len(start) > 0 && !reverse {
+		iter.Seek(start)
+	} else if hasItem {
 		iter.loadLatestItem()
-	} else {
-		iter.curItem = nil
+
+		if !iter.Valid() {
+			iter.Next()
+		}
 	}
+
+	return iter
 }
 
-func (iter *dsPrefixIter) Close(ctx context.Context) error {
-	return iter.it.Close()
+func (iter *iterator) Domain() (start []byte, end []byte) {
+	return iter.start, iter.end
 }
 
-func (iter *dsPrefixIter) loadLatestItem() {
-	curItem := iter.it.Item()
-	for iter.it.Next() {
-		if bytes.Equal(curItem.key, iter.it.Item().key) {
-			curItem = iter.it.Item()
-			continue
-		}
-		iter.it.Prev()
-		break
+func (iter *iterator) Valid() bool {
```

Comment:
suggestion: Can you document what it means to be valid for an iterator? Or rather can makes it not valid?

Reply:
I consider that out of scope here.  This is a function that implements an interface, and the documentation should exist on that interface.

None of the production code interfaces in this repo are currently documented, but that is out of scope.  I'll add a ticket though in case that gets forgotten (I don't see that as a prerequisite for integrating into Defra, but it'll be a lot easier for me to do now while everything is fresh than it would be for me/someone-else to do in a few months time).

Reply:
https://github.com/sourcenetwork/corekv/issues/30

Reply:
The file was actually lacking a type check showing that this implemented the interface, I've added one in to this PR.

Reply:
Thank you very much for flagging this Shahzad, I've found a whole bunch of issues when trying to document the public stuff!